### DISCUSSION TITLE
KnexTimeoutError: Knex: Timeout acquiring a connection.

### DISCUSSION THREAD

**Discussion by tukusejssirs** [mikro-orm/mikro-orm#2538] [Q&A] [answered]
I have tried to run a simple test of my monitoring app with 56 (virtual) machines. It runs a function every second via a cronjob and after the data is parsed (simple validation check), it is inserted into DB. Therefore there 1 insertion per machine per second, but in the end there should be 5-8 insertions per machine per second.

The problem is that after running the app for about two minutes, it throws the following error:

```ts
{repo_root}/node_modules/knex/lib/client.js:305
        convertedError = new KnexTimeoutError(
                         ^
KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at Client_PG.acquireConnection ({repo_root}/node_modules/knex/lib/client.js:305:26)
    at runNextTicks (node:internal/process/task_queues:61:5)
    at listOnTimeout (node:internal/timers:526:9)
    at processTimers (node:internal/timers:500:7)
    at Runner.ensureConnection ({repo_root}/node_modules/knex/lib/execution/runner.js:259:28)
    at Runner.run ({repo_root}/node_modules/knex/lib/execution/runner.js:30:19)
    at PostgreSqlConnection.executeQuery ({repo_root}/node_modules/@mikro-orm/core/dist/connections/Connection.js:68:25)
    at PostgreSqlConnection.execute ({repo_root}/node_modules/@mikro-orm/knex/dist/AbstractSqlConnection.js:88:21)
    at QueryBuilder.execute ({repo_root}/node_modules/@mikro-orm/knex/dist/query/QueryBuilder.js:340:21)
    at LightIndicatorService.createStatus ({repo_root}/src/light-indicator/light-indicator.service.ts:157:18)
```

I tested this on a local machine without any special load with nothing else using the DB.

I also tried to raise the default value of `max_connections` from `100` to `280` or even `400`, but it didn’t really help.

This is how I currently do the insertions:

```ts
const qb = this.orm.em.createQueryBuilder(Entity)
const status = await qb.insert(createEntityDto).withSchema(schema)
```

I was thinking to change the code that the insertion is done in memory (but Mikro-ORM should be able to work with that in-memory data, i.e. in querying the data) and `em.persistAndFlush()` all the data once a second (via a cronjob).

However, I have two questions:
1. How can I do the insertion to memory only?
2. Will `em.persistAndFlush()` flush all data in all tables or I need to run this command for each entity separately?

**Comment by B4nan** [comment, marked as answer]
Flush is always for all entities, `persistAndFlush` is just a shortcut for `persist` followed by `flush`. I am not aware of any way you could have "in memory state you could query", for that you'd have to create custom driver probably, and it sounds like a very bad idea (or better say very error prone idea) to me, but maybe I am just too scared to do such things :D

> I also tried to raise the default value of max_connections from 100 to 280 or even 400, but it didn’t really help.

I guess that is some database setting right? The error is from knex, it says the connection pool is full, so you need to adjust the knex/tarn pool settings to deal with it:

https://knexjs.org/#Installation-pooling

(MikroORM [exposes the `pool` configuration directly](https://mikro-orm.io/docs/configuration/#connection) as a top level option)

**Reply by tukusejssirs** [reply]
> Flush is always for all entities, persistAndFlush is just a shortcut for persist followed by flush. I am not aware of any way you could have "in memory state you could query", for that you'd have to create custom driver probably, and it sounds like a very bad idea (or better say very error prone idea) to me, but maybe I am just too scared to do such things :D

Okay, then forget about querying the in-memory data.

> I guess that is some database setting right?

Yes, it’s a PostgreSQL setting. :wink:

> The error is from knex, it says the connection pool is full, so you need to adjust the knex/tarn pool settings to deal with it: https://knexjs.org/#Installation-pooling

Okay. Configuring the pool settings seems to fix the issue entirely. Just one question: is it possible to change the settings at runtime? Or at least restart the Knex in order to change the pool settigs at runtime?

**Reply by tukusejssirs** [reply]
Well, the problem persists, although the errors shows up less frequently. I also noticed that there are a bit more connections (in one moment) than twice the number of insertions into the DB (i.e. right now I try to monitor 56 machines, there is a single insertion per machine per second, but the number of used connections is about `124`), but sometimes even `286`.

Therefore, I think I need to use some kind of transactions. The issue is that I need to get the data from every machine every second, that is `{number_of_machine}` insertions per second. And that is just a single feature. When other features are implemented, there would be 5-10 insertions per second (some features will be done in longer intervals, but 5-10 insertions is max). Plus we need to count the number of queries from the frontend (historical data) and CRUD operations used for user/auth/whatever management, although these won’t happen every second.

Could you help me please with dealing with this please? :pray:

**Reply by tukusejssirs** [reply]
I also noticed that sometimes MikroORM/Knex uses more connections than I set in PSQL config (`max_connections`) and MikroORM config (`pool.max`):

```sql
select max_conn,used,res_for_super,max_conn-used-res_for_super res_for_normal 
from 
  (select count(*) used from pg_stat_activity) t1,
  (select setting::int res_for_super from pg_settings where name=$$superuser_reserved_connections$$) t2,
  (select setting::int max_conn from pg_settings where name=$$max_connections$$) t3
;
 max_conn | used | res_for_super | res_for_normal 
----------+------+---------------+----------------
      280 |  285 |             3 |             -8
(1 row)
```

**Reply by B4nan** [reply]
Once again, it's not MikroORM that would do anything here, it is knex that is responsible for connection management, so I am not really in a position to help with such issues.

>  I also noticed that there are a bit more connections (in one moment) than twice the number of insertions into the DB (i.e. right now I try to monitor 56 machines, there is a single insertion per machine per second, but the number of used connections is about 124), but sometimes even 286.

That seems quite expected, each app instance will have its own connection pool, don't expect to have one connection in the pool unless you explicitly set `pool: {min: 1, max: 1}`. It's weird that it can be more than the max_connection, but as said above, that is kinda out of my domain, as knex/tarn handles this.

**Reply by tukusejssirs** [reply]
> That seems quite expected, each app instance will have its own connection pool, don't expect to have one connection in the pool unless you explicitly set `pool: {min: 1, max: 1}`. It's weird that it can be more than the max_connection, but as said above, that is kinda out of my domain, as knex/tarn handles this.

I’m a bit confused. I run a single instance of Nest.js app (unless it creates multiple instances _by itself_). Currently, I only run a single callback function in a cronjob (multiplied by the number of machines = 56 instances of that callback function). Could you elaborate your statement please?

I don’t think I really want to force the app to use a single connection.

On the max connection overflow, I’ll discuss it at Knex then. Thanks anyway. :pray: :+1:

**Reply by B4nan** [reply]
> I’m a bit confused. I run a single instance of Nest.js app (unless it creates multiple instances by itself). Currently, I only run a single callback function in a cronjob (multiplied by the number of machines = 56 instances of that callback function). Could you elaborate your statement please?

You said you have 56 virtual machines, so I'd expect that implies 56 nest app instances.

**Reply by tukusejssirs** [reply]
Sorry for the confusion. I have a single app instance that _monitors_ 56 machines. By _virtual_ I meant _non-existing [in this reality]_, because I don’t test the app with real machines, but only mock them. Sorry for describing it badly.

BTW, I tested the app overnight (~10 hours) running using `pm2`. The used connection number was usually around `54-56` (which is good enough; it seems that for some insertions it reused a connection, which is good, but it could be better) with some spikes to `133-143` lasting 1-2 minutes, which I cannot explain, because I tested it on my local machine with no other apps connecting the PSQL (not only this DB, but no other DB) and I have turned off all unnecessary programs (esp the heavy ones).

**Reply by tukusejssirs** [reply]
Just some follow-up comments based on my observations:
- It seems like the Knex pooling uses all available connections (based on `max`) or at least as many as possible. Thus I had 2~3 insertions per machine, each for different cronjob instance (in my case for different second). This suggests me that it inserted the values with 2-3 seconds delay. I solved it by lowering the `max` number of connections. I currently use `70` [`1.25 * 56`], but `56` might be better, as it still uses up all the `70` connections`, but that might not be good if the resources will be used up.
- I believe that connection ‘overuse’ can happen because in the SQL command above counts all kind of connections, including `AutoVacuumMain`, `LogicalLauncherMain`, `BgWriterMain`, `CheckpointerMain`, `WalWriterMain` (all are `wait_event`s; there might be others), and also that SQL command. Therefore I think it should be considered as false positive.

---

Another question: is it possible to modify the `max` connection number at runtime? It might be useful in my case, as the monitored machines number might change and that not all machines have same support functionalities (read: some use more cronjobs / DB insertions, some less) and we might want to modify `max` based on the machine metadata.

**Reply by B4nan** [reply]
Probably not without reconnecting, better say if you can find a way to do it with existing knex instance, you could do it here too, but I kinda dont expect that to be possible to change once instantiated.

**Reply by tukusejssirs** [reply]
FYI: runtime modification of pool configuration was already requested at Tarn (see Vincit/tarn.js#55).
